share/wake/lib/core/tree.wake

```diff
@@ -226,264 +250,307 @@ export def tfoldmap (combineFn: b => b => b) (base: b) (transformFn: a => b) (tr
         Bin _ l x r = combineFn (helper a l) (helper (transformFn x) r)
     helper base root
 
+
 export def tfold f a t = tfoldmap f a (_) t
 
+
 # Converts a Tree to a List.
-export def treeToList = tfoldr (_,_) Nil
+export def treeToList = tfoldr (_, _) Nil
+
 
 export def tappi f (Tree _ root) =
-  def helper i = match _
-    Tip = Unit
-    Bin _ l x r =
-      def ix = i + size l
-      def _ = helper i l
-      def _ = helper (ix+1) r
-      def _ = f ix x
-      Unit
-  helper 0 root
+    def helper i = match _
+        Tip = Unit
+        Bin _ l x r =
+            def ix = i + size l
+            def _ = helper i l
+            def _ = helper (ix + 1) r
+            def _ = f ix x
+            Unit
+    helper 0 root
+
 
 # Extract the i-th ranked element
 export def tat (i: Integer) (Tree _ root: Tree a): Option a =
-  def helper i = match _
-    Tip = None
-    Bin _ l x r =
-      def sizeL = size l
-      match (icmp i sizeL)
-        LT = helper i l
-        EQ = Some x
-        GT = helper (i-sizeL-1) r
-  helper i root
+    def helper i = match _
+        Tip = None
+        Bin _ l x r =
+            def sizeL = size l
+            match (icmp i sizeL)
+                LT = helper i l
+                EQ = Some x
+                GT = helper (i - sizeL - 1) r
+    helper i root
+
 
 # Split elements ranked [0,i) and [i,inf) into two trees
 export def tsplitAt (i: Integer) (Tree cmp root: Tree a): Pair (Tree a) (Tree a) =
-  def helper i = match _
-    Tip = Pair Tip Tip
-    Bin _ l x r =
-      def sizeL = size l
-      if i > sizeL
-      then match (helper (i-sizeL-1) r)
-        Pair rl rr = Pair (join3 l x rl) rr
-      else match (helper i l)
-        Pair ll lr = Pair ll (join3 lr x r)
-  match (helper i root)
-    Pair l r = Pair (Tree cmp l) (Tree cmp r)
+    def helper i = match _
+        Tip = Pair Tip Tip
+        Bin _ l x r =
+            def sizeL = size l
+            if i > sizeL then
+                match (helper (i - sizeL - 1) r)
+                    Pair rl rr = Pair (join3 l x rl) rr
+            else
+                match (helper i l)
+                    Pair ll lr = Pair ll (join3 lr x r)
```

Comment:
I think the info and primitives are there but wake-format shouldn't be doing semantic rewrites. This kind of thing should be done with a `wake-tidy`/static analyzer thing.

FWIW spinning up that tool with just this one rule wouldn't be too bad based on all the primitives from wake-format

Reply:
I agree with @V-FEXrt on the static analyzer thing. I actually had a comment for this typed up prior to Ashley's, but I apparently never submitted it. The gist of it was that I think having the formatter do rewriting, even for obviously equivalent things, is probably a slippery slope to go down. In that, probably best keeping the formatter as a pure formatter and leaving other suggestions to a static analyzer. (I.e., what Ashley said. Haha.)

If I am not mistaken, the Elixir formatter does not rewrite and only strictly formats, and [Credo](https://github.com/rrrene/credo) is used for catching things like this. (It's a little easier for a language like Elixir to have a thing like Credo since the AST is available in the language itself.)